observatory-dags/observatory/dags/telescopes/oapen_metadata.py

```diff
@@ -0,0 +1,227 @@
+
+    def __init__(self):
+        self.dag_id = OapenMetadataTelescope.DAG_ID
+        self.start_date = datetime(2018, 5, 14)
+        self.schedule_interval = '@weekly'
+        self.dataset_id = 'oapen'
+        self.merge_partition_field = 'id'
+        self.updated_date_field = 'dc.date.available'
+        self.bq_merge_days = 0
+
+        self.description = 'oapen metadata telescope'
+        self.schema_prefix = 'oapen_'
+        self.airflow_vars = [AirflowVars.DATA_PATH, AirflowVars.PROJECT_ID, AirflowVars.DATA_LOCATION,
+                             AirflowVars.DOWNLOAD_BUCKET, AirflowVars.TRANSFORM_BUCKET]
+
+        super().__init__(self.dag_id, self.start_date, self.schedule_interval, self.dataset_id,
+                         self.merge_partition_field, self.updated_date_field, self.bq_merge_days,
+                         schema_prefix=self.schema_prefix, airflow_vars=self.airflow_vars)
+
+        self.add_setup_task_chain([self.check_dependencies,
+                                   self.get_release_info])
+        self.add_task_chain([self.download,
+                             self.upload_downloaded,
+                             self.transform,
+                             self.upload_transformed,
+                             self.bq_load_partition])
+        self.add_task_chain(self.make_operators([self.bq_delete_old,
+                                                 self.bq_append_new,
+                                                 self.cleanup]))
+
+    @staticmethod
+    def make_release(**kwargs) -> 'OapenMetadataRelease':
+        # Make Release instance
+        ti: TaskInstance = kwargs['ti']
+        start_date, end_date, first_release = ti.xcom_pull(key=OapenMetadataTelescope.RELEASE_INFO,
+                                                           include_prior_dates=True)
+
+        release = OapenMetadataRelease(OapenMetadataTelescope.DAG_ID, start_date, end_date, first_release)
+        return release
+
+    def download(self, release: OapenMetadataRelease, **kwargs):
+        """ Task to download the OapenMetadataRelease release.
+        :param release: an OapenMetadataRelease instance.
+        :return: None.
+        """
+        # Download release
+        release.download()
+
+    def upload_downloaded(self, release: OapenMetadataRelease, **kwargs):
+        """ Task to upload the downloadeded OapenMetadataRelease release.
+        :param release: an OapenMetadataRelease instance.
+        :return: None.
+        """
+        # Upload each downloaded release
+        upload_files_from_list(release.download_files, release.download_bucket)
+
+    def transform(self, release: OapenMetadataRelease, **kwargs):
+        """ Task to transform the OapenMetadataRelease release.
+        :param release: an OapenMetadataRelease instance.
+        :return: None.
+        """
+        # Transform each release
+        release.transform()
+
+
+def change_keys(obj, convert, nested_fields, list_fields):
```

Comment:
This method should be broken down into smaller methods to improve readability. It's currently a bit confusing to follow.
There also seems to be a separate change_keys method in telescope_utils.

Reply:
Thanks, I agree it is a bit lengthy and messy. I took an original function 'change_keys' that is quite a bit simpler and changes the keys of a nested dictionary (this is in telescope_utils).

I might just make the changes in the actual 'oapen_metadata' branch though, I've added it here to test the template.